### PR TITLE
feat(metric-cards): new metric card summary component [MA-1375]

### PR DESCRIPTION
**Pull request by mihai-peteu**
# Summary

https://konghq.atlassian.net/browse/MA-1375
Screen Shot 2023-02-24 at 3 56 54 PM

To test, from repo root run:

```
# install dependencies for this new component
pnpm install

# opens dev preview at http://localhost:5173/
pnpm --filter "@kong-ui-public/metric-cards" run dev
```

Screen Shot 2023-02-06 at 5 31 27 PM

## PR Checklist

* [x] **Naming & Structure:** the files and package structure use the conventions outlined in the [Creating a Package docs](https://github.com/Kong/public-ui-components/blob/main/docs/creating-a-package.md).
* [x] **Tests pass:** check the output of all package unit and/or component tests.
  * [ ] If this PR is the result of a bug, test coverage was added accordingly.
* [x] **Functional:** all changes do not break existing APIs, but if so, a BREAKING CHANGE commit is in place to bump the major version.
* [x] **Conventional Commits** all commits follow the conventional commit standards [outlined in the main README](https://github.com/Kong/public-ui-components#committing-changes).
* [ ] **Docs:** includes a technically accurate README, and the docs have been updated accordingly based on the changes in this PR.

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -25,44 +25,55 @@ module.exports = {
     'no-multi-spaces': 'error',
     'no-trailing-spaces': 'error',
     'padded-blocks': 'off',
+    'no-unused-vars': 'off',
```
[the remark this answers is not in the document]
Comment (mihai-peteu) [reply .eslintrc.cjs]: i dropped it here, and added it to my specific file 👍

File: packages/analytics/metric-cards/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+// Export Vue plugin as the default
+export default {
+  // Customize Vue plugin options as desired
+  // Providing a `name` property allows for customizing the registered
+  // name of your component (useful if exporting a single component).
+  install: (app: App, options: { name?: string, [key: string]: any } = {}): void => {
+    app.component(options.name || 'MetricCards', MetricCards)
+  },
+}
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/index.ts:12]: Remove this plugin export; we would _always_ want to import these where needed for proper tree-shaking.

File: packages/analytics/metric-cards/src/types/metric-card-size.enum.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/* eslint-disable no-unused-vars */
+export enum MetricCardSize {
+  Small = 'sm',
+  Medium = 'md',
+  Large = 'lg',
+  ExtraLarge = 'xl'
+}
+/* eslint-enable no-unused-vars */
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/types/metric-card-size.enum.ts:8]: You shouldn't need the comment in here twice

File: packages/analytics/metric-cards/src/types/metric-card.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+export interface MetricsCardDef {
+  currentValue: number
+  previousValue: number,
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/types/metric-card.ts:3]: nit: no commas needed at the end of the lines (all lines)

File: packages/analytics/metric-cards/src/components/cardUtils.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+const DECIMAL_DISPLAY = 2
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/cardUtils.ts:1]: FWIW you should move this file out of the components folder; either something like `src/utilities.ts` or a composable in `src/composables/`

File: packages/analytics/metric-cards/src/styles/variables/_utilities.scss [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  color: var(--green-500, #07a88d);
+}
+
+.color-red-500 {
+  color: var(--red-500, #d44324);
+}
+
+.color-grey-500 {
+  color: var(--grey-500, #6f7787);
+}
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/styles/variables/_utilities.scss:32]: Please do not utilize these classes named the same as the utility classes we're trying to not use in this repo. Please use custom class names.

Comment (mihai-peteu) [reply packages/analytics/metric-cards/src/styles/variables/_utilities.scss:32]: Gotcha, I had a feeling you'd have an opinion on this, one way or the other :) 

This was just me being lazy, trying to get visual parity as quickly as possible.

File: packages/analytics/metric-cards/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
+{
+  "extends": "../../../tsconfig.json",
+  "compilerOptions": {
+    "baseUrl": ".",
+    "outDir": "dist",
+    "declarationDir": "dist/types",
+    "types": [
+      "node",
+      "vite/client",
+      "cypress",
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/tsconfig.json:10]: Add missing type
```suggestion
      "cypress",
      "cypress/vue",
```

File: packages/analytics/metric-cards/src/components/MetricCardContainer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+import { mount } from '@vue/test-utils'
+import MetricCardContainer from './MetricCardContainer.vue'
+
+describe('<MetricCardContainer />', () => {
+  it('renders', () => {
+    const wrapper = mount(MetricCardContainer)
+
+    expect(wrapper.isVisible()).toBe(true)
+  })
+})
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/MetricCardContainer.spec.ts:13]: @mihai-peteu you can delete this file

File: packages/analytics/metric-cards/src/components/MetricCardContainer.vue [hunk 9]
```diff
@@ -0,0 +1,97 @@
+<template>
+  <!-- The <div> tag here is just a placeholder for your component content. -->
+  <!-- We recommend wrapping your component with a unique class when possible, as shown below. -->
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/MetricCardContainer.vue:3]: nit: remove comments

File: packages/analytics/metric-cards/src/components/display/MetricCardLoadingSkeleton.vue [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    <KSkeletonBox
+      width="75"
+    />
+  </div>
+</template>
+<script setup lang="ts">
+import { KSkeletonBox } from '@kong/kongponents'
+import '@kong/kongponents/dist/style.css'
+</script>
+<style lang="scss" scoped>
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/display/MetricCardLoadingSkeleton.vue:18]: nit: can you add an empty line between blocks?

File: packages/analytics/metric-cards/src/components/display/MetricCardLoadingSkeleton.vue [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+@import "../../styles/base";
+.loading-tabs {
+  display: flex;
+  flex-direction: column;
+  padding: 16px;
+  row-gap: 12px;
+  width: 240px;
+  @media (max-width: $viewport-md) {
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/display/MetricCardLoadingSkeleton.vue:26]: issue: this rule would conflict with a min-width rule using the same breakpoint. Need to update as shown here:
```suggestion
  @media (max-width: ($viewport-md - 1px)) {
```

File: packages/analytics/metric-cards/src/components/display/MetricsCard.vue [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+<style lang="scss" scoped>
+@import "../../styles/base";
+
+.metricscard {
+  display: flex;
+  flex-direction: column;
+  max-width: 240px;
+  width: 100%;
+
+  @media (max-width: $viewport-md) {
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/display/MetricsCard.vue:161]: issue: this rule would conflict with a min-width rule using the same breakpoint. Need to update as shown here:
```suggestion
  @media (max-width: ($viewport-md - 1px)) {
```

**Comment by mihai-peteu** [comment]
@helen-tang27 Updated the "Not available" trend state.  Also, now able to differentiate between individual Card-level errors, and a Container-level error (eg: if all cards in the container contain an error):


new-state-updates

**Comment by adamdehaven** [comment]
For the error across all cards, can we use an existing pattern like `KEmptyState` or `KAlert` styled UI so it looks more consistent?

**Comment by helen-tang27** [comment]
Great !!

File: packages/analytics/metric-cards/README.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+`--kong-ui-metric-cards-trend-positive` | Positive trend font color | `#07a88d`
+`--kong-ui-metric-cards-trend-neutral` | Neutral trend font color | `#6f7787`
+
+### Install
+
+`yarn add @kong-ui-public/metric-cards`
+
+## Included Components
+
+The `MetricCardsContainer` is a parent component which contains one or more instances of `MetricCard`.
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/README.md:68]: question: Where does MetricCard live..? I'm not sure this package makes sense without it

File: packages/analytics/metric-cards/src/components/MetricCardContainer.vue [hunk 14]
```diff
@@ -0,0 +1,161 @@
+<template>
+  <div class="kong-ui-public-metric-cards">
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/MetricCardContainer.vue:2]: Should the wrapper class name be more specific to this individual component?

File: packages/analytics/metric-cards/src/components/MetricCardContainer.vue [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    </template>
+  </div>
+</template>
+
+<script setup lang="ts">
+import { computed, PropType } from 'vue'
+import approxNum from 'approximate-number'
+import { MetricCardSize, MetricsCardDef } from '../types'
+import { changePolarity, metricChange, defineIcon } from '../utilities'
+import { KIcon } from '@kong/kongponents'
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/MetricCardContainer.vue:52]: issue: you shouldn't import Kongponents, they are globally available in the host app. 
```suggestion
```

File: packages/analytics/metric-cards/src/components/MetricCardContainer.vue [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+  cardSize: {
+    type: String as PropType<MetricCardSize>,
+    required: false,
+    default: () => MetricCardSize.Large,
+  },
+})
+
+const allCardsHaveErrors = computed((): boolean => props.cards.every(val => val?.hasError === true))
+
+const calculateDelta = (curr: number, prev: number) => {
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/MetricCardContainer.vue:94]: suggestion: can you add a return type

File: packages/analytics/metric-cards/src/components/display/MetricCardLoadingSkeleton.vue [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    <KSkeletonBox
+      width="75"
+    />
+  </div>
+</template>
+
+<script setup lang="ts">
+import { KSkeletonBox } from '@kong/kongponents'
+import '@kong/kongponents/dist/style.css'
+</script>
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/display/MetricCardLoadingSkeleton.vue:15]: issue: remove kongponents imports
```suggestion
```

File: packages/analytics/metric-cards/src/components/display/MetricsCard.vue [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+          {{ metricChange }}
+        </div>
+      </div>
+    </div>
+  </div>
+</template>
+
+<script setup lang="ts">
+import { PropType } from 'vue'
+import { KIcon, KTooltip } from '@kong/kongponents'
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/display/MetricsCard.vue:55]: issue: remove kongponents imports
```suggestion
```

File: packages/analytics/metric-cards/src/components/display/MetricsCard.vue [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+import { MetricCardSize } from '../../types'
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/src/components/display/MetricsCard.vue:56]:
```suggestion
import type { MetricCardSize } from '../../types'
```

**Review by adamdehaven — CHANGES_REQUESTED** [review]

File: packages/analytics/metric-cards/src/components/MetricCardContainer.vue (hunk 14 above)
Comment (mihai-peteu) [reply packages/analytics/metric-cards/src/components/MetricCardContainer.vue:2]: Is `kong-ui-public-metric-card-container` preferable?

File: packages/analytics/metric-cards/README.md (hunk 13 above)
Comment (mihai-peteu) [reply packages/analytics/metric-cards/README.md:68]: `src/components/MetricCardContainer.vue` contains  multiple instances of `src/components/display/MetricsCard.vue`

File: packages/analytics/metric-cards/src/components/MetricCardContainer.vue (hunk 14 above)
Comment (adamdehaven) [reply packages/analytics/metric-cards/src/components/MetricCardContainer.vue:2]: sure

File: packages/analytics/metric-cards/src/components/display/MetricsCard.vue (hunk 19 above)
Comment (mihai-peteu) [reply packages/analytics/metric-cards/src/components/display/MetricsCard.vue:56]: Hmmm, if i make this change, then VS Code complains when I use it for a value comparison:

Screen Shot 2023-02-24 at 10 15 27 AM

I do this a couple of times to determine which elements are shown or hidden, eg:
```
const cardDisplayFull = [MetricCardSize.Medium, MetricCardSize.Large].includes(props.cardSize)
```

Comment (adamdehaven) [reply packages/analytics/metric-cards/src/components/display/MetricsCard.vue:56]: ah it's an enum; ignore.

File: packages/analytics/metric-cards/README.md [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+  
+
+## Features
+
+- Displays a summary panel containing various metrics
+- Supports custom color themes via [CSS variables](#css-variables)
+
+## Requirements
+
+- `vue` must be initialized in the host application
```
[the remark this answers is not in the document]
Comment (mihai-peteu) [reply packages/analytics/metric-cards/README.md:21]: Yup, added a similar mention along w/ the three Kongponents we use

```
  - `KIcon`
  - `KSkeletonBox`
  - `KTooltip`
 ```

**Comment by mihai-peteu** [comment]
@adamdehaven Another look pls. 

This PR should cover all "near future" UX feature requests for various card sizes and conditionally displayed elements (eg: hide trend and title, only show metric value, etc).  Looking forward to pulling this into khcp-ui and Dev Portal.

**Comment by mihai-peteu** [comment]
Markdown README preview:
https://github.com/Kong/public-ui-components/blob/8cd972e63312f3a8a2923a34f52e4ec0175b85f7/packages/analytics/metric-cards/README.md

**Review by adorack — DISMISSED** [review]

File: packages/analytics/metric-cards/package.json [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+    "test:unit": "cross-env FORCE_COLOR=1 vitest run",
+    "test:unit:open": "cross-env FORCE_COLOR=1 vitest --ui"
+  },
+  "peerDependencies": {
+    "@kong/kongponents": "^8.22.4",
+    "vue": "^3.2.47"
+  },
+  "devDependencies": {
+    "@kong/kongponents": "^8.22.4",
+    "approximate-number": "^2.1.0",
```
Comment (adamdehaven) [inline packages/analytics/metric-cards/package.json:46]: @mihai-peteu I believe this needs to be added as a `dependency` and removed as a `devDependency`

Comment (mihai-peteu) [reply packages/analytics/metric-cards/package.json]: Done deal! Waiting for the temp package to rebuild, will re-test in a couple of consuming apps.

Comment (mihai-peteu) [reply packages/analytics/metric-cards/package.json]: looking good still

Screenshot 2023-02-27 at 4 40 58 PM


Screenshot 2023-02-27 at 4 47 44 PM

**Review by adorack — APPROVED** [review]

**Review by adamdehaven — APPROVED** [review]
lgtm

**Pull request merged**